benches/normal.rs

```diff
@@ -0,0 +1,90 @@
+#![feature(test)]
+
+extern crate test;
+
+use fastrand::Rng;
+use test::Bencher;
+
+const SEED: u64 = 42;
+const MU: f64 = 10.0;
+const SIGMA: f64 = 3.0;
+
+#[bench]
+fn box_muller(b: &mut Bencher) {
+    let mut rng = Rng::with_seed(SEED);
+
+    b.iter(|| {
+        let mu = core::hint::black_box(MU);
+        let sigma = core::hint::black_box(SIGMA);
+
+        // https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
+        let u1 = loop {
+            let u1 = rng.f64();
+
+            if u1 > f64::EPSILON {
+                break u1;
+            }
+        };
+
+        let u2 = rng.f64();
+        let mag = sigma * (-2.0 * u1.ln()).sqrt();
+        let output = mag * (core::f64::consts::TAU * u2).cos() + mu;
+
+        core::hint::black_box(output);
+    });
+}
+
+#[bench]
+fn standard_approximation(b: &mut Bencher) {
+    let mut rng = Rng::with_seed(SEED);
+
+    b.iter(|| {
+        let mu = core::hint::black_box(MU);
+        let sigma = core::hint::black_box(SIGMA);
+
+        // http://marc-b-reynolds.github.io/distribution/2021/03/18/CheapGaussianApprox.html
+        let u1 = rng.u64(..);
+        let u2 = rng.u64(..);
+
+        let a = (u1 & 0xffffffff) as i64;
+        let b = (u1 >> 32) as i64;
+        let c = (u2 & 0xffffffff) as i64;
+        let d = (u2 >> 32) as i64;
+
+        // Magic constant.
+        let k = 3.97815e-10;
+
+        let output = k * ((a + b) - (c + d)) as f64 * sigma + mu;
+
+        core::hint::black_box(output);
+    });
+}
+
+#[bench]
+fn popcount_approximation(b: &mut Bencher) {
+    let mut rng = Rng::with_seed(SEED);
+
+    b.iter(|| {
+        let mu = core::hint::black_box(MU);
+        let sigma = core::hint::black_box(SIGMA);
+
+        // http://marc-b-reynolds.github.io/distribution/2021/03/18/CheapGaussianApprox.html
+        let u1 = rng.u64(..);
+        let bd = u1.count_ones() as i64 - 32;
+
+        let u2 = rng.u64(..);
+        let a = (u2 & 0xffffffff) as i64;
+        let b = (u2 >> 32) as i64;
+
+        let td = a - b;
+
+        let r = ((bd << 32) + td) as f64;
+
+        // Magic constant.
+        let k = 5.76917e-11;
+
+        let output = k * r * sigma + mu;
+
+        core::hint::black_box(output);
```

Comment:
To me it seems like this implementation has the least amount of magic. In addition, it could be simplified down to a single call from the RNG: calling `rng.u128(..)` and then splitting that into the two `u64`s. Right now this is just implemented as two calls to `rng.u64(..)`, but:

- Theoretically this could be made faster with a different RNG implementation with optimized `u128`.
- It would avoid more than one call to thread-local storage.

Reply:
Yeah, I looked at the current implementation of `rng.u128(..)` and saw two `u64`s, but didn't realize the advantages that you stated here. I will change that.

According to the linked post, this method also has lower error from Box-Muller reference than the other approximation, although it's a little bit slower, at least on my computer. I will also look for some resources and try to explain what's happening in the comments.

Reply:
I changed benchmarks and the implementation to use u128 and added comments in the implementation (it is still quite vague for people without stronger background in probability theory, like me, but it at least contains some terms that can be searched for on the internet).

I did these changes in separate commits to simplify review, but I will eventually squash it into one.